File: src/main/java/org/kambanaria/knots/examples/DoubleUserKnot.java
package org.kambanaria.knots.examples;

import java.util.List;
import org.kambanaria.knots.microservices_models.Account;
import org.kambanaria.knots.microservices_models.Bank;
import org.kambanaria.knots.microservices_models.User;
import static org.kambanaria.knots.services.StaticRegistry.*;

public class DoubleUserKnot {

    private final int userId;
    private final int bankId;
    private User user;
    private Bank bank;
    private List<Account> accounts;

    public DoubleUserKnot(int userId, int bankId) {
        this.userId = userId;
        this.bankId = bankId;
    }

    public int getUserId() {
        return userId;
    }

    public User getUser() {
        if (user == null) {
            user = USER_SERVICE.getUserById(userId);
        }
        return user;
    }

    public List<Account> getAccounts() {
        if (accounts == null) {
            accounts = ACCOUNT_SERVICE.getAccountsByUserIdBankId(userId, bankId);
        }
        return accounts;
    }

    public Bank getBank() {
        if (bank == null) {
            bank = BANK_SERVICE.getBankById(bankId);
        }
        return bank;
    }
}

File: src/main/python/double_user_knot.py
from services import UserService as USER_SERVICE
from services import AccountService as ACCOUNT_SERVICE
from services import BankService as BANK_SERVICE


class DoubleUserKnot(object):
    def __init__(self, user_id, bank_id):
        self.user_id = user_id
        self.bank_id = bank_id
        self._cached_user = None
        self._cached_bank = None
        self._cached_accounts = None

    @property
    def user(self):
        if self._cached_user is None:
            self._cached_user = USER_SERVICE.get_user_by_id(self.user_id)
        return self._cached_user

    @property
    def bank(self):
        if self._cached_bank is None:
            self._cached_bank = BANK_SERVICE.get_bank_by_id(self.bank_id)
        return self._cached_bank

    @property
    def accounts(self):
        if self._cached_accounts is None:
            self._cached_accounts = ACCOUNT_SERVICE.get_accounts_by_user_id_bank_id(
                self.user_id, self.bank_id)
        return self._cached_accounts
